### ISSUE TITLE
Record total packages and versions in stats

### ISSUE THREAD

**Issue by thomashoneyman** [purescript/registry-dev#237]
Currently, the `Stats` module outputs statistics with this format:

```json
Number of successful packages: 1369
Number of failed packages: 474
Number of successful versions: 9565
Number of failed versions: 1694
Failures by error:
  manifestError: 1378 occurrences (371 packages / 1378 versions)
    missingLicense: 928 occurrences (241 packages / 928 versions)
    badDependencyVersions: 283 occurrences (104 packages / 283 versions)
    badLicense: 179 occurrences (45 packages / 179 versions)
    badVersion: 48 occurrences (32 packages / 48 versions)
  nonRegistryDependencies: 246 occurrences (47 packages / 246 versions)
  missingBowerfile: 64 occurrences (34 packages / 64 versions)
  noReleases: 32 occurrences (32 packages / 0 versions)
  malformedBowerJson: 5 occurrences (4 packages / 5 versions)
  malformedPackageName: 1 occurrences (1 packages / 1 versions)
```

However, it would also be useful to know the _total_ number of packages and versions we attempted to work with. For example, these statistics list 1,369 "successful" packages and 474 "failed" packages, but we only have something like 1500-1600 packages in total; some packages are present in both the success and failure maps because _some_, but not all, of their versions succeeded.

I think the output of the first few lines ought to be more like:

```
Packages: 1500 total (1369 with successes, 474 with failures)
Versions: 11000 total (9565 successful, 1694 failed)
```

where the total packages won't be a sum of its constituents because some packages will be in both maps, but where the total _versions_ should be the sum of its constituents because a version can't be failed _and_ successful.

We could also in the future be more granular and segment packages with all successes, all failures, and a mixture of both, but that's a larger task.

**Comment by thomashoneyman** [comment]
We can get the total number of packages with a `Map.size` of this value:

https://github.com/purescript/registry/blob/28ffb8412ccd09130a81c1aecc27c2a59b7b6674/ci/src/Registry/Scripts/BowerImport.purs#L78

and we have all releases that we're going to try and fetch in the `releaseIndex` -- this corresponds to "versions" we're going to attempt to parse:

https://github.com/purescript/registry/blob/28ffb8412ccd09130a81c1aecc27c2a59b7b6674/ci/src/Registry/Scripts/BowerImport.purs#L94

**Comment by thomashoneyman** [comment]
In fact, as discussed in the registry channel, I think we really ought to list packages that _totally_ failed (only present in the failures map), packages that _partially_ failed (present in both maps), and packages that _totally_ succeeded (only present in success map, the registry index)

**Comment by mikesol** [comment]
On it :+1:

**Comment by mikesol** [comment]
> In fact, as discussed in the registry channel, I think we really ought to list packages that _totally_ failed (only present in the failures map), packages that _partially_ failed (present in both maps), and packages that _totally_ succeeded (only present in success map, the registry index)

Will do this tomorrow :+1:

**Comment by mikesol** [comment]
Are the indices in the maps the same (ie are they all identifiers of packages)? If so then the sorting & printing is a lot easier. Also, if that's the case, it may be worth it to have a newtype for the key that's used in all maps (ie `PackageIdentifier`) & to get rid of the polymorphism in the map types so that it's more transparent what's going on.

**Comment by thomashoneyman** [comment]
Yes, by the end of the process you’ve got a map of PackageName to SemVer to Manifest; everything is typed at that point. Feel free to specialize!

**Comment by mikesol** [comment]
At the risk of yak-shaving, I'm tidying up the types and trying to figure out what each `String` represents so I know if I'm reporting the right thing or not. The first port of call is here:

```purescript
readRegistryFile :: FilePath -> Aff (Map RawPackageName String)
readRegistryFile source = do
  registryFile <- readJsonFile ("../" <> source)
  case registryFile of
    Left err -> do
      let decodeError = "Decoding " <> source <> "failed with error:\n\n" <> Json.printJsonDecodeError err
      throwError $ Aff.error decodeError
    Right packages -> do
      let toPackagesArray = Object.toArrayWithKey \k -> Tuple (RawPackageName $ stripPureScriptPrefix k)
      pure $ Map.fromFoldable $ toPackagesArray packages
```

`RawPackageName` totally makes sense, but I'm not sure what the value `String` represents here as I don't know the spec of the files being parsed. If you can include an example file in this issue and/or and example of what the parsed map looks like, that'd help! I tried running `GITHUB_TOKEN=foo npx spago run -m Registry.API` but it needs a bit more elaborate setup than that, so I wasn't able to get the code to this function & extract anything useful. Thanks in advance for your help!

**Comment by thomashoneyman** [comment]
> At the risk of yak-shaving,

Please yak-shave away!

> I tried running `GITHUB_TOKEN=foo npx spago run -m Registry.API` but it needs a bit more elaborate setup than that, so I wasn't able to get the code to this function & extract anything useful. Thanks in advance for your help!

I think you should copy the existing `.env.example` file over with your GitHub token, and then run `spago run -m Registry.Scripts.BowerImport`. The `Registry.API` module is specific for CI, and doesn't have anything to do with the Bower import script.

> `RawPackageName` totally makes sense, but I'm not sure what the value `String` represents here as I don't know the spec of the files being parsed.

A "registry file" is either https://github.com/purescript/registry/blob/master/bower-packages.json or https://github.com/purescript/registry/blob/master/new-packages.json, ie. a mapping of a package name to a GitHub repository. Since they're all going to be URLs, please feel free to use `Http.URL` to type them! (We import Affjax as `Http`, if I recall correctly.)

**Comment by mikesol** [comment]
> Please yak-shave away!
> 

![](https://media3.giphy.com/media/Gd0A7bNRh5UNW/giphy.gif)

I added a `PackageURL` newtype.

**Comment by thomashoneyman** [comment]
> I added a `PackageURL` newtype.

😆 As with all good yak-shaving, what do you think of using the `Http.URL` type synonym from Affjax instead of a dedicated `PackageURL` newtype? I know we newtyped all the map keys for the processing steps, so that we'd get type errors if we merged maps incorrectly, but I'd rather stay away from having lots of modules in the registry depend on this processing script (for example, `Foreign.GitHub` shouldn't depend on the import script modules).

If the `URL` type synonym brings enough clarity in your opinion would you be open to switching?

**Comment by mikesol** [comment]
> > I added a `PackageURL` newtype.
> 
> If the `URL` type synonym brings enough clarity in your opinion would you be open to switching?

It depends on the topology of the yak to be shaved.

Here, the URL is not only http protocol - it sometimes is, but it is also sometimes other stuff (this I say judging from the intent of the original author):

```purescript
parseRepo :: PackageURL -> Either Parser.ParseError Address
parseRepo = unwrap >>> Parser.runParser do
  void $ Parse.string "https://github.com/"
    <|> Parse.string "git://github.com/"
    <|> Parse.string "git@github.com:"
```

So it seems that `Http.URL` would not be the best fit.It's not a biggie, though - if that makes it more clear, we can for sure use it!

**Comment by thomashoneyman** [comment]
Good point. Perhaps just a URL type synonym exported from the Registry.Prelude module, then? Or we can do PackageURL — but I think it should be in the prelude if we’re using it outside the legacy import script.

**Comment by f-f** [comment]
This was fixed in #239 right?
